File: apiwdbe/DlgWdbeUsrMngkeypair.cpp
/**
	* \file DlgWdbeUsrMngkeypair.cpp
	* API code for job DlgWdbeUsrMngkeypair (implementation)
	* \author Alexander Wirthmueller
	* \date created: 23 Aug 2020
	* \date modified: 23 Aug 2020
	*/

#include "DlgWdbeUsrMngkeypair.h"

using namespace std;
using namespace Sbecore;
using namespace Xmlio;

/******************************************************************************
 class DlgWdbeUsrMngkeypair::VecVDo
 ******************************************************************************/

uint DlgWdbeUsrMngkeypair::VecVDo::getIx(
			const string& sref
		) {
	string s = StrMod::lc(sref);

	if (s == "detbutdelclick") return DETBUTDELCLICK;
	if (s == "detbutgenclick") return DETBUTGENCLICK;
	if (s == "butdneclick") return BUTDNECLICK;

	return(0);
};

string DlgWdbeUsrMngkeypair::VecVDo::getSref(
			const uint ix
		) {
	if (ix == DETBUTDELCLICK) return("DetButDelClick");
	if (ix == DETBUTGENCLICK) return("DetButGenClick");
	if (ix == BUTDNECLICK) return("ButDneClick");

	return("");
};

/******************************************************************************
 class DlgWdbeUsrMngkeypair::VecVSge
 ******************************************************************************/

uint DlgWdbeUsrMngkeypair::VecVSge::getIx(
			const string& sref
		) {
	string s = StrMod::lc(sref);

	if (s == "nf") return NF;
	if (s == "found") return FOUND;

	return(0);
};

string DlgWdbeUsrMngkeypair::VecVSge::getSref(
			const uint ix
		) {
	if (ix == NF) return("nf");
	if (ix == FOUND) return("found");

	return("");
};

/******************************************************************************
 class DlgWdbeUsrMngkeypair::ContInf
 ******************************************************************************/

DlgWdbeUsrMngkeypair::ContInf::ContInf(
			const uint numFSge
			, const string& DetTxtSte
			, const string& DetDld
		) :
			Block()
		{
	this->numFSge = numFSge;
	this->DetTxtSte = DetTxtSte;
	this->DetDld = DetDld;

	mask = {NUMFSGE, DETTXTSTE, DETDLD};
};

bool DlgWdbeUsrMngkeypair::ContInf::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "ContInfDlgWdbeUsrMngkeypair");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "ContitemInfDlgWdbeUsrMngkeypair";

	if (basefound) {
		if (extractUintAttrUclc(docctx, basexpath, itemtag, "Ci", "sref", "numFSge", numFSge)) add(NUMFSGE);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ci", "sref", "DetTxtSte", DetTxtSte)) add(DETTXTSTE);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ci", "sref", "DetDld", DetDld)) add(DETDLD);
	};

	return basefound;
};

set<uint> DlgWdbeUsrMngkeypair::ContInf::comm(
			const ContInf* comp
		) {
	set<uint> items;

	if (numFSge == comp->numFSge) insert(items, NUMFSGE);
	if (DetTxtSte == comp->DetTxtSte) insert(items, DETTXTSTE);
	if (DetDld == comp->DetDld) insert(items, DETDLD);

	return(items);
};

set<uint> DlgWdbeUsrMngkeypair::ContInf::diff(
			const ContInf* comp
		) {
	set<uint> commitems;
	set<uint> diffitems;

	commitems = comm(comp);

	diffitems = {NUMFSGE, DETTXTSTE, DETDLD};
	for (auto it = commitems.begin(); it != commitems.end(); it++) diffitems.erase(*it);

	return(diffitems);
};

/******************************************************************************
 class DlgWdbeUsrMngkeypair::StatApp
 ******************************************************************************/

DlgWdbeUsrMngkeypair::StatApp::StatApp(
			const string& shortMenu
		) :
			Block()
		{
	this->shortMenu = shortMenu;

	mask = {SHORTMENU};
};

bool DlgWdbeUsrMngkeypair::StatApp::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "StatAppDlgWdbeUsrMngkeypair");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "StatitemAppDlgWdbeUsrMngkeypair";

	if (basefound) {
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "shortMenu", shortMenu)) add(SHORTMENU);
	};

	return basefound;
};

set<uint> DlgWdbeUsrMngkeypair::StatApp::comm(
			const StatApp* comp
		) {
	set<uint> items;

	if (shortMenu == comp->shortMenu) insert(items, SHORTMENU);

	return(items);
};

set<uint> DlgWdbeUsrMngkeypair::StatApp::diff(
			const StatApp* comp
		) {
	set<uint> commitems;
	set<uint> diffitems;

	commitems = comm(comp);

	diffitems = {SHORTMENU};
	for (auto it = commitems.begin(); it != commitems.end(); it++) diffitems.erase(*it);

	return(diffitems);
};

/******************************************************************************
 class DlgWdbeUsrMngkeypair::StatShr
 ******************************************************************************/

DlgWdbeUsrMngkeypair::StatShr::StatShr(
			const bool DetButDelAvail
			, const bool DetButGenAvail
			, const bool DetDldActive
		) :
			Block()
		{
	this->DetButDelAvail = DetButDelAvail;
	this->DetButGenAvail = DetButGenAvail;
	this->DetDldActive = DetDldActive;

	mask = {DETBUTDELAVAIL, DETBUTGENAVAIL, DETDLDACTIVE};
};

bool DlgWdbeUsrMngkeypair::StatShr::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "StatShrDlgWdbeUsrMngkeypair");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "StatitemShrDlgWdbeUsrMngkeypair";

	if (basefound) {
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "DetButDelAvail", DetButDelAvail)) add(DETBUTDELAVAIL);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "DetButGenAvail", DetButGenAvail)) add(DETBUTGENAVAIL);
		if (extractBoolAttrUclc(docctx, basexpath, itemtag, "Si", "sref", "DetDldActive", DetDldActive)) add(DETDLDACTIVE);
	};

	return basefound;
};

set<uint> DlgWdbeUsrMngkeypair::StatShr::comm(
			const StatShr* comp
		) {
	set<uint> items;

	if (DetButDelAvail == comp->DetButDelAvail) insert(items, DETBUTDELAVAIL);
	if (DetButGenAvail == comp->DetButGenAvail) insert(items, DETBUTGENAVAIL);
	if (DetDldActive == comp->DetDldActive) insert(items, DETDLDACTIVE);

	return(items);
};

set<uint> DlgWdbeUsrMngkeypair::StatShr::diff(
			const StatShr* comp
		) {
	set<uint> commitems;
	set<uint> diffitems;

	commitems = comm(comp);

	diffitems = {DETBUTDELAVAIL, DETBUTGENAVAIL, DETDLDACTIVE};
	for (auto it = commitems.begin(); it != commitems.end(); it++) diffitems.erase(*it);

	return(diffitems);
};

/******************************************************************************
 class DlgWdbeUsrMngkeypair::Tag
 ******************************************************************************/

DlgWdbeUsrMngkeypair::Tag::Tag(
			const string& Cpt
			, const string& DetCptSte
			, const string& DetButDel
			, const string& DetButGen
			, const string& DetDld
			, const string& ButDne
		) :
			Block()
		{
	this->Cpt = Cpt;
	this->DetCptSte = DetCptSte;
	this->DetButDel = DetButDel;
	this->DetButGen = DetButGen;
	this->DetDld = DetDld;
	this->ButDne = ButDne;

	mask = {CPT, DETCPTSTE, DETBUTDEL, DETBUTGEN, DETDLD, BUTDNE};
};

bool DlgWdbeUsrMngkeypair::Tag::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "TagDlgWdbeUsrMngkeypair");
	else
		basefound = checkXPath(docctx, basexpath);

	string itemtag = "TagitemDlgWdbeUsrMngkeypair";

	if (basefound) {
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ti", "sref", "Cpt", Cpt)) add(CPT);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ti", "sref", "DetCptSte", DetCptSte)) add(DETCPTSTE);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ti", "sref", "DetButDel", DetButDel)) add(DETBUTDEL);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ti", "sref", "DetButGen", DetButGen)) add(DETBUTGEN);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ti", "sref", "DetDld", DetDld)) add(DETDLD);
		if (extractStringAttrUclc(docctx, basexpath, itemtag, "Ti", "sref", "ButDne", ButDne)) add(BUTDNE);
	};

	return basefound;
};

/******************************************************************************
 class DlgWdbeUsrMngkeypair::DpchAppDo
 ******************************************************************************/

DlgWdbeUsrMngkeypair::DpchAppDo::DpchAppDo(
			const string& scrJref
			, const uint ixVDo
			, const set<uint>& mask
		) :
			DpchAppWdbe(VecWdbeVDpch::DPCHAPPDLGWDBEUSRMNGKEYPAIRDO, scrJref)
		{
	if (find(mask, ALL)) this->mask = {SCRJREF, IXVDO};
	else this->mask = mask;

	this->ixVDo = ixVDo;
};

string DlgWdbeUsrMngkeypair::DpchAppDo::getSrefsMask() {
	vector<string> ss;
	string srefs;

	if (has(SCRJREF)) ss.push_back("scrJref");
	if (has(IXVDO)) ss.push_back("ixVDo");

	StrMod::vectorToString(ss, srefs);

	return(srefs);
};

void DlgWdbeUsrMngkeypair::DpchAppDo::writeXML(
			xmlTextWriter* wr
		) {
	xmlTextWriterStartElement(wr, BAD_CAST "DpchAppDlgWdbeUsrMngkeypairDo");
	xmlTextWriterWriteAttribute(wr, BAD_CAST "xmlns", BAD_CAST "http://www.mpsitech.com/wdbe");
		if (has(SCRJREF)) writeString(wr, "scrJref", scrJref);
		if (has(IXVDO)) writeString(wr, "srefIxVDo", VecVDo::getSref(ixVDo));
	xmlTextWriterEndElement(wr);
};

/******************************************************************************
 class DlgWdbeUsrMngkeypair::DpchEngData
 ******************************************************************************/

DlgWdbeUsrMngkeypair::DpchEngData::DpchEngData() :
			DpchEngWdbe(VecWdbeVDpch::DPCHENGDLGWDBEUSRMNGKEYPAIRDATA)
		{
	feedFSge.tag = "FeedFSge";
};

string DlgWdbeUsrMngkeypair::DpchEngData::getSrefsMask() {
	vector<string> ss;
	string srefs;

	if (has(SCRJREF)) ss.push_back("scrJref");
	if (has(CONTINF)) ss.push_back("continf");
	if (has(FEEDFSGE)) ss.push_back("feedFSge");
	if (has(STATAPP)) ss.push_back("statapp");
	if (has(STATSHR)) ss.push_back("statshr");
	if (has(TAG)) ss.push_back("tag");

	StrMod::vectorToString(ss, srefs);

	return(srefs);
};

void DlgWdbeUsrMngkeypair::DpchEngData::readXML(
			xmlXPathContext* docctx
			, string basexpath
			, bool addbasetag
		) {
	clear();

	bool basefound;

	if (addbasetag)
		basefound = checkUclcXPaths(docctx, basexpath, basexpath, "DpchEngDlgWdbeUsrMngkeypairData");
	else
		basefound = checkXPath(docctx, basexpath);

	if (basefound) {
		if (extractStringUclc(docctx, basexpath, "scrJref", "", scrJref)) add(SCRJREF);
		if (continf.readXML(docctx, basexpath, true)) add(CONTINF);
		if (feedFSge.readXML(docctx, basexpath, true)) add(FEEDFSGE);
		if (statapp.readXML(docctx, basexpath, true)) add(STATAPP);
		if (statshr.readXML(docctx, basexpath, true)) add(STATSHR);
		if (tag.readXML(docctx, basexpath, true)) add(TAG);
	} else {
		continf = ContInf();
		feedFSge.clear();
		statapp = StatApp();
		statshr = StatShr();
		tag = Tag();
	};
};

File: japiwdbe/DlgWdbeUsrMngkeypair.java
/**
  * \file DlgWdbeUsrMngkeypair.java
  * Java API code for job DlgWdbeUsrMngkeypair
	* \copyright (C) 2018-2020 MPSI Technologies GmbH
	* \author Alexander Wirthmueller (auto-generation)
	* \date created: 5 Dec 2020
	*/
// IP header --- ABOVE

package apiwdbe;

import java.util.*;
import org.w3c.dom.*;
import sbecore.*;

public class DlgWdbeUsrMngkeypair {
	/**
		* VecVDo (full: VecVDlgWdbeUsrMngkeypairDo)
		*/
	public static class VecVDo {

		public static final int DETBUTDELCLICK = 1;
		public static final int DETBUTGENCLICK = 2;
		public static final int BUTDNECLICK = 3;

		public static int getIx(
					String sref
				) {
			String s = sref.toLowerCase();

			if (s.equals("detbutdelclick")) return DETBUTDELCLICK;
			if (s.equals("detbutgenclick")) return DETBUTGENCLICK;
			if (s.equals("butdneclick")) return BUTDNECLICK;

			return 0;
		};

		public static String getSref(
					int ix
				) {
			if (ix == DETBUTDELCLICK) return("DetButDelClick");
			if (ix == DETBUTGENCLICK) return("DetButGenClick");
			if (ix == BUTDNECLICK) return("ButDneClick");

			return "";
		};

	};

	/**
		* VecVSge (full: VecVDlgWdbeUsrMngkeypairSge)
		*/
	public static class VecVSge {

		public static final int NF = 1;
		public static final int FOUND = 2;

		public static int getIx(
					String sref
				) {
			String s = sref.toLowerCase();

			if (s.equals("nf")) return NF;
			if (s.equals("found")) return FOUND;

			return 0;
		};

		public static String getSref(
					int ix
				) {
			if (ix == NF) return("nf");
			if (ix == FOUND) return("found");

			return "";
		};

	};

	/**
	  * ContInf (full: ContInfDlgWdbeUsrMngkeypair)
	  */
	public class ContInf extends Block {

		public static final int NUMFSGE = 1;
		public static final int DETTXTSTE = 2;
		public static final int DETDLD = 3;

		public ContInf(
					int numFSge
					, String DetTxtSte
					, String DetDld
				) {
			this.numFSge = numFSge;
			this.DetTxtSte = DetTxtSte;
			this.DetDld = DetDld;

			mask = new HashSet<Integer>(Arrays.asList(NUMFSGE, DETTXTSTE, DETDLD));
		};

		public int numFSge;
		public String DetTxtSte;
		public String DetDld;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "ContInfDlgWdbeUsrMngkeypair");

			String itemtag = "ContitemInfDlgWdbeUsrMngkeypair";

			if (Xmlio.checkXPath(doc, basexpath)) {
				numFSge = Xmlio.extractIntegerAttrUclc(doc, basexpath, itemtag, "Ci", "sref", "numFSge", mask, NUMFSGE);
				DetTxtSte = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ci", "sref", "DetTxtSte", mask, DETTXTSTE);
				DetDld = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ci", "sref", "DetDld", mask, DETDLD);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					ContInf comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (numFSge == comp.numFSge) items.add(NUMFSGE);
			if (DetTxtSte.equals(comp.DetTxtSte)) items.add(DETTXTSTE);
			if (DetDld.equals(comp.DetDld)) items.add(DETDLD);

			return(items);
		};

		public HashSet<Integer> diff(
					ContInf comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(NUMFSGE, DETTXTSTE, DETDLD));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
	  * StatApp (full: StatAppDlgWdbeUsrMngkeypair)
	  */
	public class StatApp extends Block {

		public static final int SHORTMENU = 1;

		public StatApp(
					String shortMenu
				) {
			this.shortMenu = shortMenu;

			mask = new HashSet<Integer>(Arrays.asList(SHORTMENU));
		};

		public String shortMenu;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "StatAppDlgWdbeUsrMngkeypair");

			String itemtag = "StatitemAppDlgWdbeUsrMngkeypair";

			if (Xmlio.checkXPath(doc, basexpath)) {
				shortMenu = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Si", "sref", "shortMenu", mask, SHORTMENU);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					StatApp comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (shortMenu.equals(comp.shortMenu)) items.add(SHORTMENU);

			return(items);
		};

		public HashSet<Integer> diff(
					StatApp comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(SHORTMENU));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
	  * StatShr (full: StatShrDlgWdbeUsrMngkeypair)
	  */
	public class StatShr extends Block {

		public static final int DETBUTDELAVAIL = 1;
		public static final int DETBUTGENAVAIL = 2;
		public static final int DETDLDACTIVE = 3;

		public StatShr(
					boolean DetButDelAvail
					, boolean DetButGenAvail
					, boolean DetDldActive
				) {
			this.DetButDelAvail = DetButDelAvail;
			this.DetButGenAvail = DetButGenAvail;
			this.DetDldActive = DetDldActive;

			mask = new HashSet<Integer>(Arrays.asList(DETBUTDELAVAIL, DETBUTGENAVAIL, DETDLDACTIVE));
		};

		public boolean DetButDelAvail;
		public boolean DetButGenAvail;
		public boolean DetDldActive;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "StatShrDlgWdbeUsrMngkeypair");

			String itemtag = "StatitemShrDlgWdbeUsrMngkeypair";

			if (Xmlio.checkXPath(doc, basexpath)) {
				DetButDelAvail = Xmlio.extractBooleanAttrUclc(doc, basexpath, itemtag, "Si", "sref", "DetButDelAvail", mask, DETBUTDELAVAIL);
				DetButGenAvail = Xmlio.extractBooleanAttrUclc(doc, basexpath, itemtag, "Si", "sref", "DetButGenAvail", mask, DETBUTGENAVAIL);
				DetDldActive = Xmlio.extractBooleanAttrUclc(doc, basexpath, itemtag, "Si", "sref", "DetDldActive", mask, DETDLDACTIVE);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					StatShr comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (DetButDelAvail == comp.DetButDelAvail) items.add(DETBUTDELAVAIL);
			if (DetButGenAvail == comp.DetButGenAvail) items.add(DETBUTGENAVAIL);
			if (DetDldActive == comp.DetDldActive) items.add(DETDLDACTIVE);

			return(items);
		};

		public HashSet<Integer> diff(
					StatShr comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(DETBUTDELAVAIL, DETBUTGENAVAIL, DETDLDACTIVE));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
	  * Tag (full: TagDlgWdbeUsrMngkeypair)
	  */
	public class Tag extends Block {

		public static final int CPT = 1;
		public static final int DETCPTSTE = 2;
		public static final int DETBUTDEL = 3;
		public static final int DETBUTGEN = 4;
		public static final int DETDLD = 5;
		public static final int BUTDNE = 6;

		public Tag(
					String Cpt
					, String DetCptSte
					, String DetButDel
					, String DetButGen
					, String DetDld
					, String ButDne
				) {
			this.Cpt = Cpt;
			this.DetCptSte = DetCptSte;
			this.DetButDel = DetButDel;
			this.DetButGen = DetButGen;
			this.DetDld = DetDld;
			this.ButDne = ButDne;

			mask = new HashSet<Integer>(Arrays.asList(CPT, DETCPTSTE, DETBUTDEL, DETBUTGEN, DETDLD, BUTDNE));
		};

		public String Cpt;
		public String DetCptSte;
		public String DetButDel;
		public String DetButGen;
		public String DetDld;
		public String ButDne;

		public boolean readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "TagDlgWdbeUsrMngkeypair");

			String itemtag = "TagitemDlgWdbeUsrMngkeypair";

			if (Xmlio.checkXPath(doc, basexpath)) {
				Cpt = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "Cpt", mask, CPT);
				DetCptSte = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "DetCptSte", mask, DETCPTSTE);
				DetButDel = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "DetButDel", mask, DETBUTDEL);
				DetButGen = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "DetButGen", mask, DETBUTGEN);
				DetDld = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "DetDld", mask, DETDLD);
				ButDne = Xmlio.extractStringAttrUclc(doc, basexpath, itemtag, "Ti", "sref", "ButDne", mask, BUTDNE);

				return true;
			};

			return false;
		};

		public HashSet<Integer> comm(
					Tag comp
				) {
			HashSet<Integer> items = new HashSet<Integer>();

			if (Cpt.equals(comp.Cpt)) items.add(CPT);
			if (DetCptSte.equals(comp.DetCptSte)) items.add(DETCPTSTE);
			if (DetButDel.equals(comp.DetButDel)) items.add(DETBUTDEL);
			if (DetButGen.equals(comp.DetButGen)) items.add(DETBUTGEN);
			if (DetDld.equals(comp.DetDld)) items.add(DETDLD);
			if (ButDne.equals(comp.ButDne)) items.add(BUTDNE);

			return(items);
		};

		public HashSet<Integer> diff(
					Tag comp
				) {
			HashSet<Integer> commitems;
			HashSet<Integer> diffitems;

			commitems = comm(comp);

			diffitems = new HashSet<Integer>(Arrays.asList(CPT, DETCPTSTE, DETBUTDEL, DETBUTGEN, DETDLD, BUTDNE));
			for (Integer ci: commitems) diffitems.remove(ci);

			return(diffitems);
		};

	};

	/**
		* DpchAppDo (full: DpchAppDlgWdbeUsrMngkeypairDo)
		*/
	public class DpchAppDo extends DpchAppWdbe {

		public static final int SCRJREF = 1;
		public static final int IXVDO = 2;
		public static final int ALL = 3;

		public DpchAppDo(
					String scrJref
					, int ixVDo
					, Integer[] mask
				) {
			super(VecWdbeVDpch.DPCHAPPDLGWDBEUSRMNGKEYPAIRDO, scrJref);

			this.mask = new HashSet<Integer>(Arrays.asList(mask));

			for (Integer i: mask)
				if (i == ALL) {
					this.mask = new HashSet<Integer>(Arrays.asList(SCRJREF, IXVDO));
					break;
				};

			this.ixVDo = ixVDo;
		};

		public int ixVDo;

		public String getSrefsMask() {
			ArrayList<String> ss = new ArrayList<String>();

			if (has(SCRJREF)) ss.add("scrJref");
			if (has(IXVDO)) ss.add("ixVDo");

			return StrMod.vectorToString(ss, ';');
		};

		public void writeXML(
					Document doc
					, Element sup
				) {
			Element el = doc.createElement("DpchAppDlgWdbeUsrMngkeypairDo");

			if (sup == null) doc.appendChild(el);
			else sup.appendChild(el);

			el.setAttribute("xmlns", "http://www.mpsitech.com");

			if (has(SCRJREF)) Xmlio.writeString(doc, el, "scrJref", scrJref);
			if (has(IXVDO)) Xmlio.writeString(doc, el, "srefIxVDo", VecVDo.getSref(ixVDo));
		};

	};

	/**
		* DpchEngData (full: DpchEngDlgWdbeUsrMngkeypairData)
		*/
	public class DpchEngData extends DpchEngWdbe {

		public static final int SCRJREF = 1;
		public static final int CONTINF = 2;
		public static final int FEEDFSGE = 3;
		public static final int STATAPP = 4;
		public static final int STATSHR = 5;
		public static final int TAG = 6;

		public DpchEngData() {
			super(VecWdbeVDpch.DPCHENGDLGWDBEUSRMNGKEYPAIRDATA);

			continf = new ContInf(0, "", "");
			feedFSge = new Feed("FeedFSge");
			statapp = new StatApp("");
			statshr = new StatShr(false, false, false);
			tag = new Tag("", "", "", "", "", "");
		};

		public ContInf continf;
		public Feed feedFSge;
		public StatApp statapp;
		public StatShr statshr;
		public Tag tag;

		public String getSrefsMask() {
			ArrayList<String> ss = new ArrayList<String>();

			if (has(SCRJREF)) ss.add("scrJref");
			if (has(CONTINF)) ss.add("continf");
			if (has(FEEDFSGE)) ss.add("feedFSge");
			if (has(STATAPP)) ss.add("statapp");
			if (has(STATSHR)) ss.add("statshr");
			if (has(TAG)) ss.add("tag");

			return StrMod.vectorToString(ss, ';');
		};

		public void readXML(
					Document doc
					, String basexpath
					, boolean addbasetag
				) {

			clear();

			if (addbasetag) basexpath = Xmlio.checkUclcXPaths(doc, basexpath, "DpchEngDlgWdbeUsrMngkeypairData");

			if (Xmlio.checkXPath(doc, basexpath)) {
				scrJref = Xmlio.extractStringUclc(doc, basexpath, "scrJref", "", mask, SCRJREF);
				if (continf.readXML(doc, basexpath, true)) add(CONTINF);
				if (feedFSge.readXML(doc, basexpath, true)) add(FEEDFSGE);
				if (statapp.readXML(doc, basexpath, true)) add(STATAPP);
				if (statshr.readXML(doc, basexpath, true)) add(STATSHR);
				if (tag.readXML(doc, basexpath, true)) add(TAG);
			} else {
				scrJref = "";
				continf = new ContInf(0, "", "");
				feedFSge = new Feed("FeedFSge");
				statapp = new StatApp("");
				statshr = new StatShr(false, false, false);
				tag = new Tag("", "", "", "", "", "");
			};
		};

	};

};

